vivisect/tools/graphutil.py

```diff
@@ -658,16 +701,47 @@ def preRouteGraphDown(graph, fromva):
     if fromnode == None:
         raise Exception("fromva not in graph 0x%x" % fromva)
 
+    nwlist = graph.getHierNodeWeights()
     todo = [ graph.getNode(fromnode) ]
     while todo:
-        curnodeva, curnode = todo.pop()
-        graph.getNodeProps(curnodeva)['up'] = True
-        for eid, fr, to, einfo in graph.getRefsFrom((curnodeva,curnode)):
-            if graph.getNodeProps(to).get('up') == True:
+        curnode = todo.pop()
+        curnodeva, curninfo = curnode
+        graph.setNodeProp(curnode, mark, True)
+        for eid, fr, to, einfo in graph.getRefsFrom(curnode):
+            if graph.getNodeProps(to).get(mark) == True:
                 continue
-            #raw_input("try next")
+            if not loop and nwlist.get(fr) >= nwlist.get(to):
+                continue
+
             todo.append(graph.getNode(to))
 
+def clearMarkDown(graph, fromva, loop=False, mark='up'):
```

Comment:
very specific clearing of tags (properties).  very useful for marking subsets of graph-flow and then clearing a different subset of graph-flow, almost like a ven diagram but for graph-flow